### PR TITLE
Add sample estimation in SortAndSampleKernel

### PR DESCRIPTION
**Pull request by jeanp413**

### CODE REVIEW

**Comment by kharoc** [comment]
1 RAL

========================================================
FAILED TESTS                                                                                                                                                                                                
========================================================                                                                                                                                                    
                        TestGroup InputType Result                                           TestId
0  Bindable Alias                  orc       Fail   TEST_09                                        
1  Date                            orc       Fail   TEST_01                                        
2  Simple Distribution From Local  orc       Fail   TEST_22,TEST_43,TEST_46,TEST_49                
3  TPCH Queries                    orc       Fail   TEST_03,TEST_05,TEST-08,TEST_10,TEST_14,TEST_18
4  Timestamp                       orc       Fail   TEST_02                                        
*******************************************************************************
TOTAL SUMMARY for test suite: 
PASSED: 1261/1274
FAILED: 13/1274
CRASH: 0/1274
TOTAL: 1274
MAX DELTA: 216.0
*******************************************************************************
Aggregations without group by:   Start Mem: 427.875   End Mem: 541.875   Diff: 114.0
Coalesce:   Start Mem: 541.875   End Mem: 495.875   Diff: -46.0
Column Basis:   Start Mem: 495.875   End Mem: 495.875   Diff: 0.0
Common Table Expressions:   Start Mem: 495.875   End Mem: 577.875   Diff: 82.0
Count without group by:   Start Mem: 577.875   End Mem: 579.875   Diff: 2.0
Date:   Start Mem: 579.875   End Mem: 605.875   Diff: 26.0
Timestamp:   Start Mem: 605.875   End Mem: 651.875   Diff: 46.0
Full outer join:   Start Mem: 651.875   End Mem: 651.875   Diff: 0.0
Group by:   Start Mem: 651.875   End Mem: 643.875   Diff: -8.0
Group by without aggregations:   Start Mem: 643.875   End Mem: 625.875   Diff: -18.0
Inner join:   Start Mem: 625.875   End Mem: 695.875   Diff: 70.0
Left outer join:   Start Mem: 695.875   End Mem: 677.875   Diff: -18.0
Non-EquiJoin Queries:   Start Mem: 677.875   End Mem: 789.875   Diff: 112.0
Nested Queries:   Start Mem: 789.875   End Mem: 831.875   Diff: 42.0
Order by:   Start Mem: 831.875   End Mem: 797.875   Diff: -34.0
Predicates With Nulls:   Start Mem: 797.875   End Mem: 863.875   Diff: 66.0
Simple String:   Start Mem: 863.875   End Mem: 851.875   Diff: -12.0
Unary ops:   Start Mem: 913.875   End Mem: 1013.875   Diff: 100.0
Unify Tables:   Start Mem: 1013.875   End Mem: 961.875   Diff: -52.0
Union:   Start Mem: 961.875   End Mem: 1031.875   Diff: 70.0
Limit:   Start Mem: 1031.875   End Mem: 931.875   Diff: -100.0
Where clause:   Start Mem: 931.875   End Mem: 1049.875   Diff: 118.0
Bindable Alias:   Start Mem: 1049.875   End Mem: 1185.875   Diff: 136.0
Boolean:   Start Mem: 1185.875   End Mem: 1123.875   Diff: -62.0
Case:   Start Mem: 1123.875   End Mem: 1153.875   Diff: 30.0
Cast:   Start Mem: 1153.875   End Mem: 1245.875   Diff: 92.0
Concat:   Start Mem: 1245.875   End Mem: 1327.875   Diff: 82.0
Literal:   Start Mem: 1327.875   End Mem: 1309.875   Diff: -18.0
Dir:   Start Mem: 1309.875   End Mem: 1525.875   Diff: 216.0
File System Local:   Start Mem: 1525.875   End Mem: 1439.875   Diff: -86.0
File System S3:   Start Mem: 1439.875   End Mem: 1437.875   Diff: -2.0
File System Google Storage :   Start Mem: 1437.875   End Mem: 1411.875   Diff: -26.0
Like:   Start Mem: 1411.875   End Mem: 1521.875   Diff: 110.0
Substring:   Start Mem: 1633.875   End Mem: 1743.875   Diff: 110.0
Wild Card:   Start Mem: 1743.875   End Mem: 1761.875   Diff: 18.0
TPCH Queries:   Start Mem: 1761.875   End Mem: 1891.875   Diff: 130.0
Round:   Start Mem: 1891.875   End Mem: 1953.875   Diff: 62.0

2 Worker's

========================================================
FAILED TESTS                                                                                                                                                                                                
========================================================                                                                                                                                                    
                        TestGroup InputType Result                                           TestId
0  Bindable Alias                  orc       Fail   TEST_09                                        
1  Date                            orc       Fail   TEST_01                                        
2  Simple Distribution From Local  orc       Fail   TEST_22,TEST_43,TEST_46,TEST_49                
3  TPCH Queries                    orc       Fail   TEST_03,TEST_05,TEST-08,TEST_10,TEST_14,TEST_18
4  Timestamp                       orc       Fail   TEST_02                                        
*******************************************************************************
TOTAL SUMMARY for test suite: 
PASSED: 1500/1513
FAILED: 13/1513
CRASH: 0/1513
TOTAL: 1513
MAX DELTA: 471.0
*******************************************************************************
Aggregations without group by:   Start Mem: 844.875   End Mem: 1315.875   Diff: 471.0
Coalesce:   Start Mem: 1315.875   End Mem: 1397.875   Diff: 82.0
Common Table Expressions:   Start Mem: 1397.875   End Mem: 1429.875   Diff: 32.0
Count without group by:   Start Mem: 1429.875   End Mem: 1477.875   Diff: 48.0
Date:   Start Mem: 1477.875   End Mem: 1463.875   Diff: -14.0
Timestamp:   Start Mem: 1463.875   End Mem: 1461.875   Diff: -2.0
Full outer join:   Start Mem: 1461.875   End Mem: 1437.875   Diff: -24.0
Group by:   Start Mem: 1437.875   End Mem: 1477.875   Diff: 40.0
Group by without aggregations:   Start Mem: 1477.875   End Mem: 1549.875   Diff: 72.0
Inner join:   Start Mem: 1549.875   End Mem: 1535.875   Diff: -14.0
Left outer join:   Start Mem: 1535.875   End Mem: 1623.875   Diff: 88.0
Non-EquiJoin Queries:   Start Mem: 1623.875   End Mem: 1733.875   Diff: 110.0
Nested Queries:   Start Mem: 1733.875   End Mem: 1731.875   Diff: -2.0
Order by:   Start Mem: 1731.875   End Mem: 1849.875   Diff: 118.0
Predicates With Nulls:   Start Mem: 1849.875   End Mem: 2015.875   Diff: 166.0
Simple String:   Start Mem: 2015.875   End Mem: 2047.875   Diff: 32.0
Unary ops:   Start Mem: 2195.875   End Mem: 2261.875   Diff: 66.0
Unify Tables:   Start Mem: 2261.875   End Mem: 2241.875   Diff: -20.0
Union:   Start Mem: 2241.875   End Mem: 2211.875   Diff: -30.0
Limit:   Start Mem: 2211.875   End Mem: 2029.875   Diff: -182.0
Where clause:   Start Mem: 2029.875   End Mem: 2021.875   Diff: -8.0
Bindable Alias:   Start Mem: 2021.875   End Mem: 2131.875   Diff: 110.0
Boolean:   Start Mem: 2131.875   End Mem: 2191.875   Diff: 60.0
Case:   Start Mem: 2191.875   End Mem: 2245.875   Diff: 54.0
Cast:   Start Mem: 2245.875   End Mem: 2349.875   Diff: 104.0
Concat:   Start Mem: 2349.875   End Mem: 2395.875   Diff: 46.0
Literal:   Start Mem: 2395.875   End Mem: 2453.875   Diff: 58.0
Dir:   Start Mem: 2453.875   End Mem: 2823.875   Diff: 370.0
File System Local:   Start Mem: 2823.875   End Mem: 2693.875   Diff: -130.0
File System S3:   Start Mem: 2693.875   End Mem: 2737.875   Diff: 44.0
File System Google Storage :   Start Mem: 2737.875   End Mem: 2807.875   Diff: 70.0
Like:   Start Mem: 2807.875   End Mem: 2971.875   Diff: 164.0
Substring:   Start Mem: 3197.875   End Mem: 3345.875   Diff: 148.0
Wild Card:   Start Mem: 3345.875   End Mem: 3515.875   Diff: 170.0
TPCH Queries:   Start Mem: 3515.875   End Mem: 3645.875   Diff: 130.0
Round:   Start Mem: 3645.875   End Mem: 3753.875   Diff: 108.0

File: engine/src/execution_graph/logic_controllers/BatchOrderByProcessing.h [hunk 1]
```diff
@@ -150,10 +150,30 @@ class SortAndSampleKernel : public kernel {
...
+			localTotalNumRows, avg_bytes_per_row, this->expression, this->context.get());
+		this->add_to_output_cache(std::move(partitionPlan), "output_b");
+	}
 	
 	virtual kstatus run() {
 		CodeTimer timer;
 
+		bool try_num_rows_estimation = true;
+		bool estimate_samples = false;
+		uint64_t num_sample_rows_estimate = 0;
```
Comment (williamBlazing) [inline engine/src/execution_graph/logic_controllers/BatchOrderByProcessing.h:167]: this should be population_to_sample

File: engine/src/execution_graph/logic_controllers/BatchOrderByProcessing.h [hunk 2]
```diff
@@ -150,10 +150,30 @@ class SortAndSampleKernel : public kernel {
...
+		uint64_t sum_sample_rows = 0;
```
Comment (williamBlazing) [inline engine/src/execution_graph/logic_controllers/BatchOrderByProcessing.h:168]: this should be population_sampled

File: engine/src/execution_graph/logic_controllers/BatchOrderByProcessing.h [hunk 3]
```diff
@@ -169,6 +189,21 @@ class SortAndSampleKernel : public kernel {
 				sampledTables.push_back(std::move(sampledTable));
 				localTotalNumRows += batch->view().num_rows();
 				localTotalBytes += batch->sizeInBytes();
+
+				// Try samples estimation
+				if(try_num_rows_estimation) {
+					uint64_t num_rows_estimate;
+					std::tie(estimate_samples, num_rows_estimate) = this->query_graph->get_estimated_input_rows_to_cache(this->get_id(), std::to_string(this->get_id()));
+					num_sample_rows_estimate = static_cast<uint64_t>(num_sample_rows_estimate * ORDER_BY_SAMPLES_RATIO);
```
Comment (williamBlazing) [inline engine/src/execution_graph/logic_controllers/BatchOrderByProcessing.h:197]: this was logically incorrect. should be
population_to_sample = static_cast<uint64_t>(num_rows_estimate * ORDER_BY_SAMPLES_RATIO);

File: engine/src/execution_graph/logic_controllers/BatchOrderByProcessing.h [hunk 4]
```diff
@@ -169,6 +189,21 @@ class SortAndSampleKernel : public kernel {
...
+					try_num_rows_estimation = false;
+				}
+				sum_sample_rows += sampledTables.back()->num_rows();
```
Comment (williamBlazing) [inline engine/src/execution_graph/logic_controllers/BatchOrderByProcessing.h:200]: We are not trying to collect how many samples, but how big is the population we sampled. So this should be:
population_sampled += batch->num_rows();

File: engine/src/execution_graph/logic_controllers/BatchOrderByProcessing.h [hunk 5]
```diff
@@ -169,6 +189,21 @@ class SortAndSampleKernel : public kernel {
...
+				if (estimate_samples && sum_sample_rows > num_sample_rows_estimate)	{
```
Comment (williamBlazing) [inline engine/src/execution_graph/logic_controllers/BatchOrderByProcessing.h:201]: update variable names:
if (estimate_samples && population_sampled > population_to_sample)

**Pull request merged**